### ISSUE TITLE
Nullable warning for passing known non-null string to non-nullable ref arg

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#34874]
**Version Used**: 
3.1.0-beta1-19172-05+edd2de88fb3e84a097fb30b4070e0f219f624e40

**Steps to Reproduce**:
```C#
#nullable enable
public class Program
{
    public static void Main()
    {
        string? argument = "hello";
        Method(ref argument);
    }

    static void Method(ref string parameter) { }
}
```
This warns:
```
Argument of type 'string?' cannot be used for parameter 'parameter' of type 'string' in 'void Program.Method(ref string parameter)' due to differences in the nullability of reference types.
```
Why is this warning needed?  We know that `argument` is non-null, so it won't violate any expectations of `Method`, and there's nothing `Method` could set `parameter` to that would violate expectations of the caller upon `Method's` return.

cc: @cston, @dotnet/nullablefc

**Comment by cston** [comment]
cc @jcouv

**Comment by jcouv** [comment]
Tagging @gafter for comment while I'm away.

**Comment by gafter** [comment]
This is a known issue.

**Comment by jcouv** [comment]
I don't remember exactly when/where, but I have a vague recollection this was an explicit decision (types must match).
I'll chat with Neal today to confirm desired behavior. We can easily remove this check.

**Comment by jcouv** [comment]
From discussion with Mads, it does feel like we should just check the safety of inbound and outbound assignments, rather than checking that the declared types match exactly.
In this scenario, after the `Method(ref argument);` call, we should know that `argument` is not-null.

**Comment by RikkiGibson** [comment]
@jcouv it appears the scenario in the issue description has the expected behavior now. Should this issue be closed?

**Comment by jcouv** [comment]
@RikkiGibson The logic in `VisitArgumentConversionAndInboundAssignmentsAndPreConditions` still looks like the old description ("checking that the declared types match exactly") so we probably still need to take a pass and confirm we have the desired behavior.

```
    case RefKind.Ref:
                    if (!argumentNoConversion.IsSuppressed)
                    {
                        var lvalueResultType = result.LValueType;
                        if (IsNullabilityMismatch(lvalueResultType.Type, parameterType.Type))
                        {
                            // declared types must match, ignoring top-level nullability
                            ReportNullabilityMismatchInRefArgument(argumentNoConversion, argumentType: lvalueResultType.Type, parameter, parameterType.Type);
                        }
                        else
                        {
                            // types match, but state would let a null in
                            ReportNullableAssignmentIfNecessary(argumentNoConversion, ApplyLValueAnnotations(parameterType, parameterAnnotations), resultType, useLegacyWarnings: false);
                            // If the parameter has annotations, we perform an additional check for nullable value types
                            CheckDisallowedNullAssignment(resultType, parameterAnnotations, argumentNoConversion.Syntax.Location);
                        }
                    }
```

**Comment by RikkiGibson** [comment]
The comment `// declared types must match, ignoring top-level nullability` seems to indicate that we intentionally accept the scenario in the issue description.
